lib/widgets/action_sheet.dart

```diff
@@ -400,3 +403,69 @@ class ShareButton extends MessageActionSheetMenuItemButton {
     }
   }
 }
+
+class MarkAsUnread extends MessageActionSheetMenuItemButton {
+  MarkAsUnread({
+    super.key,
+    required super.message,
+    required super.messageListContext,
+  });
+
+  @override IconData get icon => Icons.mark_chat_unread_outlined;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return zulipLocalizations.actionSheetMarkAsUnread;
+  }
+
+  @override void onPressed(BuildContext context) async {
+    Navigator.of(context).pop();
+    final zulipLocalizations = ZulipLocalizations.of(messageListContext);
+
+    try {
+      final store = PerAccountStoreWidget.of(messageListContext);
+      final narrow = SendableNarrow.ofMessage(message, selfUserId: store.selfUserId);
```

Comment:
This will always return a `TopicNarrow` no matter which narrow the message is shown on. This results in marking the messages unread just in a single topic narrow even if we are in a stream narrow.

Reply:
Right! I wonder though is there another alternative other that receiving the narrow in showMessageActionSheet arguments?

Reply:
I think passing that Narrow down is a proper option for now.

Reply:
Passing down a Narrow would work but would be a bit annoying because it'd have to be passed down through to the MessageWithPossibleSender that calls showMessageActionSheet.

I've just written up and sent PR #811, which provides a way to do this similar to what these action sheets already do to get the compose box controller (for quote-and-reply etc.). So @Khader-1 please go ahead and stack this PR atop that one, and then you can use that.

Reply:
Thanks @gnprice! I've just done that in the new revision. PTAL @sm-sayedi.